### PR TITLE
Fix #1788 Add pre-commit to lint unit tests

### LINKED ISSUE #1788: pre-commit verifier linter
# Issue Type

- Feature request

# Linting molecule tests

I'd like to have a pre-commit linter to verify my ansible role test suite.

### PR DESCRIPTION
**Pull request by dmlb2000**
#### PR Type

- Bugfix Pull Request

Fix #1788 

This adds a pre-commit verifier linter to run pre-commit on the
molecule unit tests.

Signed-off-by: David Brown <dmlb2000@gmail.com>

### CODE REVIEW

**Comment by decentral1se** [comment]
Thanks for this! Think we need some clarifying comments over at https://github.com/ansible/molecule/issues/1788 before moving ahead with this.

File: molecule/verifier/lint/precommit.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+              remove-tabs:
+
+    Test file linting can be disabled by setting `enabled` to False.
+
+    .. code-block:: yaml
+
+        verifier:
+          name: testinfra
+          lint:
+            name: precommit
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]:
```suggestion
            name: pre-commit
```

File: molecule/verifier/lint/precommit.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            enabled: False
+
+    Environment variables can be passed to lint.
+
+    .. code-block:: yaml
+
+        verifier:
+          name: testinfra
+          lint:
+            name: precommit
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]:
```suggestion
            name: pre-commit
```

File: test/unit/verifier/lint/test_precommit.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+import pytest
+import sh
+
+from molecule import config
+from molecule.verifier.lint import precommit
+
+
+@pytest.fixture
+def _patched_get_tests(mocker):
```
Comment (webknjaz) [inline test/unit/verifier/lint/test_precommit.py]: could you please use buitin fixture for this?
```suggestion
def _patched_get_tests(mokeypatch):
```

**Comment by dmlb2000** [comment]
@webknjaz not sure what the deal is with the linting the output seems a bit weird to me... thoughts?

**Comment by decentral1se** [comment]
@dmlb2000 you can run `tox -e format` and just accept what `yapf` thinks is right :)

File: molecule/verifier/lint/precommit.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
+#  Copyright (c) 2015-2018 Cisco Systems, Inc.
```
Comment (decentral1se) [inline molecule/verifier/lint/precommit.py]: Think this needs to be `Copyright (c) 2018‒2019 Red Hat, Inc.`. See other more recently edited files.

File: molecule/verifier/lint/precommit.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    .. code-block:: yaml
+
+        verifier:
+          name: testinfra
+          lint:
+            name: pre-commit
+            options:
+              remove-tabs:
+
+    Test file linting can be disabled by setting `enabled` to False.
```
Comment (decentral1se) [inline molecule/verifier/lint/precommit.py]: We'll need double backquotes for code samples. 

See http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#inline-markup.

File: test/unit/verifier/lint/test_precommit.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
+#  Copyright (c) 2015-2018 Cisco Systems, Inc.
```
Comment (decentral1se) [inline test/unit/verifier/lint/test_precommit.py]: Same as above. Needs to be marked as `... Red Hat ...`.

File: test/unit/verifier/lint/test_precommit.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+
+def test_enabled_property(_instance):
+    assert _instance.enabled
+
+
+@pytest.mark.parametrize(
+    'config_instance', ['_verifier_lint_section_data'], indirect=True)
+def test_options_property(_instance):
+    x = {
```
Comment (decentral1se) [inline test/unit/verifier/lint/test_precommit.py]: This is totally optional BUT if you want to use more useful naming, that would be great :)

**Comment by decentral1se** [comment]
I'm open to persuasion but in the interest of a more stable v2.20, I'm pushing this to v2.21.

/cc @themr0c

File: molecule/verifier/lint/precommit.py (hunk 4 above)
Comment (dmlb2000) [reply molecule/verifier/lint/precommit.py]: A lot of this code was directly copied from the `flake8.py` in the same directory and modified from there. What suggestions should I apply to `flake8.py` as well?

**Comment by dmlb2000** [comment]
> Nice one @dmlb2000!
> 
> Can we also get a functional test in some Docker based scenario (the only one we have working reliably right now and can be run easily be a developer). Something in https://github.com/ansible/molecule/tree/master/test/scenarios oughta do it!

Is there some pipeline that's running this? I'm having trouble finding this in `.travis.yml` or `tox.ini`

File: molecule/verifier/lint/precommit.py (hunk 5 above)
Comment (dmlb2000) [reply molecule/verifier/lint/precommit.py]: Should be done, and fixed in `flake8.py` as well.

File: test/unit/verifier/lint/test_precommit.py (hunk 6 above)
Comment (dmlb2000) [reply test/unit/verifier/lint/test_precommit.py]: Done!

File: test/unit/verifier/lint/test_precommit.py (hunk 7 above)
Comment (dmlb2000) [reply test/unit/verifier/lint/test_precommit.py]: I changed the options to be more realistic for pre-commit. (disable color output).

File: molecule/verifier/lint/precommit.py (hunk 4 above)
Comment (webknjaz) [reply molecule/verifier/lint/precommit.py]: flake8 was developed before the acquisition of Molecule by Ansible Project/Red Hat. So it only applies to new code. But It shouldn't be `2015-`, because you didn't write it 4 years ago.

File: molecule/verifier/lint/precommit.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
+#  Copyright (c) 2018-2019 Red Hat, Inc.
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]:
```suggestion
#  Copyright (c) 2019 Red Hat, Inc.
```
This wasn't written in 2018.

File: test/unit/verifier/lint/test_precommit.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
+#  Copyright (c) 2018-2019 Red Hat, Inc.
```
Comment (webknjaz) [inline test/unit/verifier/lint/test_precommit.py]:
```suggestion
#  Copyright (c) 2019 Red Hat, Inc.
```

File: molecule/verifier/lint/precommit.py (hunk 4 above)
Comment (dmlb2000) [reply molecule/verifier/lint/precommit.py]: Okay, thanks for the fix I'm going to call this one resolved.

File: molecule/verifier/lint/precommit.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+from molecule import logger
+from molecule import util
+from molecule.verifier.lint import base
+
+LOG = logger.get_logger(__name__)
+
+
+class PreCommit(base.Base):
+    """
+    `PreCommit`_ is not the default verifier linter.
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: I think we don't need a link in the docstring title + please put it in the first line.
Also, following PEP 257, it should be a sentence explaining what the class represents. Currently it looks like a wikipedia summary.

File: molecule/verifier/lint/precommit.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+from molecule import logger
+from molecule import util
+from molecule.verifier.lint import base
+
+LOG = logger.get_logger(__name__)
+
+
+class PreCommit(base.Base):
+    """
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]:
```suggestion
    """Pre-commit tool verifier wrapper.

```

File: molecule/verifier/lint/precommit.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+            name: pre-commit
+            env:
+              FOO: bar
+
+    .. _`Pre-Commit`: https://pre-commit.com/
+    """
+
+    def __init__(self, config):
+        """
+        Sets up the requirements to execute `pre-commit` and returns None.
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: It should be in imperative mood.

```suggestion
        Set up the requirements to execute `pre-commit` tool.
```

"returns None" doesn't make sense because it's an initializer it's a common knowledge that it cannot return anything other than `None`.

File: molecule/verifier/lint/precommit.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+        :param config: An instance of a Molecule config.
+        :return: None
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: this is probably not needed for initializer

File: molecule/verifier/lint/precommit.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+            _out=LOG.out,
+            _err=LOG.error)
+
+    def execute(self):
+        if not self.enabled:
+            msg = 'Skipping, verifier_lint is disabled.'
+            LOG.warn(msg)
+            return
+
+        if not len(self._tests) > 0:
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: This is unpythonic. I cannot even say what this check does right away with the double negation... The proper way to do this is
```suggestion
        if not self._tests:
```

File: molecule/verifier/lint/precommit.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+            msg = 'Skipping, no tests found.'
+            LOG.warn(msg)
+            return
+
+        if self._precommit_command is None:
+            self.bake()
+
+        msg = 'Executing Pre-Commit on files found in {}/...'.format(
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: The tool name is lower-case (check the website and official texts)
```suggestion
        msg = 'Executing pre-commit on files found in {}/...'.format(
```

File: molecule/verifier/lint/precommit.py [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        except sh.ErrorReturnCode as e:
+            util.sysexit(e.exit_code)
+
+    def _get_tests(self):
+        """
+        Walk the verifier's directory for tests and returns a list.
+
+        :return: list
+        """
+        return [
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: What are you doing?
```suggestion
        return list(util.os_walk(self._config.verifier.directory, 'test_*.py'))
```

File: molecule/verifier/lint/precommit.py (hunk 10 above)
Comment (dmlb2000) [reply molecule/verifier/lint/precommit.py]: I'd like to run pep257 on a lot of the code but that would require making a lot of changes across all the code at that point...

Comment (dmlb2000) [reply molecule/verifier/lint/precommit.py]: But this should be resolved otherwise..

File: test/unit/verifier/lint/test_precommit.py (hunk 3 above)
Comment (dmlb2000) [reply test/unit/verifier/lint/test_precommit.py]: I'm assuming you mean `monkeypatch` here? I am going to need more of an example here to make the conversion there's only two references to `monkeypatch` in the unit tests and the are manipulating environment variables not methods in a class. I'm pretty used to dealing with `mock` and `patch` to fake class methods seems to work okay for me.

File: molecule/verifier/lint/flake8.py [hunk 17]
```diff
@@ -35,7 +35,7 @@ class Flake8(base.Base):
 
     `Flake8`_ is a linter for python files.
 
-    Additional options can be passed to `flake8` through the options
+    Additional options can be passed to ``flake8`` through the options
```
Comment (webknjaz) [inline molecule/verifier/lint/flake8.py]: This module is unrelated, please submit this change as a separate PR

File: molecule/verifier/lint/precommit.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+from molecule import logger
+from molecule import util
+from molecule.verifier.lint import base
+
+LOG = logger.get_logger(__name__)
+
+
+class PreCommit(base.Base):
+    """
+    Pre-commit linter class to execute pre-commit tool for linting files.
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: Title should fit in 72 characters and be in the first line if possible as per PEP257

File: molecule/verifier/lint/precommit.py [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+        try:
+            util.run_command(self._precommit_command, debug=self._config.debug)
+            msg = 'Lint completed successfully.'
+            LOG.success(msg)
+        except sh.ErrorReturnCode as e:
+            util.sysexit(e.exit_code)
+
+    def _get_tests(self):
+        """
+        Walk the verifier's directory for tests and returns a list.
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: Docstring title shouldn't attempt to explain implementation details. It should mention the intention/high-level effect, nothing more. Everything else should go into a separate description paragraph.

File: test/unit/verifier/lint/test_precommit.py (hunk 3 above)
Comment (webknjaz) [reply test/unit/verifier/lint/test_precommit.py]: mock has certain bugs with unpleasant side-effects + there's not much point in having multiple tools doing the same. Why use extra dependency if there's a built-in thing which is officially recommend to use by pytest devs? That's why I'm trying to encourage folks to use `monkeypatch` now.

Comment (webknjaz) [reply test/unit/verifier/lint/test_precommit.py]: https://docs.pytest.org/en/latest/monkeypatch.html

File: test/unit/verifier/lint/test_precommit.py [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+import pytest
+import sh
+
+from molecule import config
+from molecule.verifier.lint import precommit
+
+
+@pytest.fixture
+def _patched_get_tests(mocker):
+    m = mocker.patch('molecule.verifier.lint.precommit.PreCommit._get_tests')
```
Comment (webknjaz) [inline test/unit/verifier/lint/test_precommit.py]:
```suggestion
    monkeypatch.setattr('molecule.verifier.lint.precommit.PreCommit._get_tests', lambda: ['test1', 'test2', 'test3'])
    yield
```

File: test/unit/verifier/lint/test_precommit.py [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+    m.return_value = ['test1', 'test2', 'test3']
```
Comment (webknjaz) [inline test/unit/verifier/lint/test_precommit.py]:
```suggestion
```

File: test/unit/verifier/lint/test_precommit.py [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+
+    return m
```
Comment (webknjaz) [inline test/unit/verifier/lint/test_precommit.py]:
```suggestion
```

File: test/unit/verifier/lint/test_precommit.py [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+
+from molecule import config
+from molecule.verifier.lint import precommit
+
+
+@pytest.fixture
+def _patched_get_tests(mocker):
+    m = mocker.patch('molecule.verifier.lint.precommit.PreCommit._get_tests')
+    m.return_value = ['test1', 'test2', 'test3']
+
```
Comment (webknjaz) [inline test/unit/verifier/lint/test_precommit.py]:
```suggestion
```

File: test/unit/verifier/lint/test_precommit.py (hunk 7 above)
Comment (webknjaz) [reply test/unit/verifier/lint/test_precommit.py]: It was about naming X with something more meaningful. What does it refer to?

File: test/unit/verifier/lint/test_precommit.py (hunk 3 above)
Comment (dmlb2000) [reply test/unit/verifier/lint/test_precommit.py]: Just FYI `mock` is now part of core https://docs.python.org/3/library/unittest.mock.html so there's no extra dependency anymore.

Could you elaborate on the unpleasant side-effects?

File: test/unit/verifier/lint/test_precommit.py (hunk 20 above)
Comment (dmlb2000) [reply test/unit/verifier/lint/test_precommit.py]: Should be fixed now.

File: test/unit/verifier/lint/test_precommit.py (hunk 7 above)
Comment (dmlb2000) [reply test/unit/verifier/lint/test_precommit.py]: This one is fixed too.

File: test/unit/verifier/lint/test_precommit.py (hunk 3 above)
Comment (webknjaz) [reply test/unit/verifier/lint/test_precommit.py]: You use `pytest-mock`, which provides `mocker` fixture. It's not exactly stdlib mock.

Side-effects is that it breaks pytest itself sometimes. Longer explanation: https://github.com/ansible/molecule/pull/1533#discussion_r260881603

File: test/scenarios/verifier/.pre-commit-config.yaml [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+      - id: check-symlinks
+      - id: detect-private-key
+  - repo: local
+    hooks:
+      - id: pylint
+        name: pylint
+        entry: python -m pylint
+        language: system
+        types: [python]
+  - repo: git://github.com/Lucas-C/pre-commit-hooks
```
Comment (decentral1se) [inline test/scenarios/verifier/.pre-commit-config.yaml]: Can we avoid an external repository? Unsure if all these checks are related to that but can we have instead a 'hello, world' check or something that just smokes out that it runs anything successfully. Sorry for my continued lack of pre-commit knowledge :)

Comment (decentral1se) [reply test/scenarios/verifier/.pre-commit-config.yaml]: Is this related to https://github.com/ansible/molecule/issues/1788#issuecomment-468806847?

File: molecule/verifier/lint/precommit.py [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+#  all copies or substantial portions of the Software.
+#
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+#  DEALINGS IN THE SOFTWARE.
+"""Linter module for pre-commit related code."""
```
Comment (decentral1se) [inline molecule/verifier/lint/precommit.py]: Nitpick warning: new lines like in other modules?

File: test/scenarios/verifier/.pre-commit-config.yaml (hunk 24 above)
Comment (dmlb2000) [reply test/scenarios/verifier/.pre-commit-config.yaml]: Some of it is just my own communication with you guys, to show you the extent of pre-commit and what it can do. I agree that for the fixture a local hook that just runs something simple would be preferred. However, I would think that a pre-commit config like this could easily take the place of the lint and format checks in your `tox.ini` at the root of the repository.

Comment (dmlb2000) [reply test/scenarios/verifier/.pre-commit-config.yaml]: I'll just run the flake8 hook since that's what you guys run through tox

**Comment by dmlb2000** [comment]
@decentral1se @webknjaz if either of you two could re-roll the travis pipeline, I'm not sure that error is my fault.

**Comment by decentral1se** [comment]
Re-running, that build is quite flaky! Let's see ...

File: molecule/verifier/lint/precommit.py [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+class PreCommit(base.Base):
+    """
+    Pre-commit linter class.
+
+    This class is used to lint files by executing the pre-commit
+    command line tool  linting files.
+
+    `PreCommit`_ is not the default verifier linter.
+
+    `PreCommit`_ is a linter for python files and more.
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: It should match the reference declaration below
```suggestion
    `Pre-Commit`_ is a linter for python files and more.
```
https://github.com/ansible/molecule/pull/1789/files#diff-46d6f9714ab4084355923b5649e1817bR77

File: molecule/verifier/lint/precommit.py [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+
+
+class PreCommit(base.Base):
+    """
+    Pre-commit linter class.
+
+    This class is used to lint files by executing the pre-commit
+    command line tool  linting files.
+
+    `PreCommit`_ is not the default verifier linter.
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]:
```suggestion
    `Pre-Commit`_ is not the default verifier linter.
```

File: molecule/verifier/lint/precommit.py [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+from molecule import logger
+from molecule import util
+from molecule.verifier.lint import base
+
+LOG = logger.get_logger(__name__)
+
+
+class PreCommit(base.Base):
+    """
+    Pre-commit linter class.
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: up ^

File: molecule/verifier/lint/precommit.py [hunk 29]
```diff
@@ -0,0 +1,144 @@
...
+
+    This class is used to lint files by executing the pre-commit
+    command line tool  linting files.
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: I believe there's a preposition missing in the place of double whitespace.

```suggestion
    command line tool for linting files.
```

File: molecule/verifier/lint/precommit.py [hunk 30]
```diff
@@ -0,0 +1,144 @@
...
+        :param config: An instance of a Molecule config.
+        """
+        super(PreCommit, self).__init__(config)
+        self._precommit_command = None
+        if config:
+            self._tests = self._get_tests()
+
+    @property
+    def default_options(self):
+        """Property of the default options for pre-commit."""
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: Excuse me, what? It doesn't look descriptive to me. It's obvious that it's a property just because it's decorated `@property` two lines above.
Let's create a more helpful docsrting. It should explain what the property actually represents, not how it's implemented.

```suggestion
        """Default options for pre-commit tool runtime."""
```

File: molecule/verifier/lint/precommit.py [hunk 31]
```diff
@@ -0,0 +1,144 @@
...
+        return {}
+
+    @property
+    def default_env(self):
+        """Property of the default environment for pre-commit."""
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: Same:
```suggestion
        """Default environment variables for pre-commit tool runtime."""
```

File: molecule/verifier/lint/precommit.py [hunk 32]
```diff
@@ -0,0 +1,144 @@
...
+        return util.merge_dicts(os.environ.copy(), self._config.env)
+
+    def bake(self):
+        """Bake a `pre-commit` command so it's ready to execute."""
```
Comment (webknjaz) [inline molecule/verifier/lint/precommit.py]: The title shouldn't explain the motivation, it should declare effect/action. Also, RST requires double backticks for inline code highlighting.

```suggestion
        """Bake a ready to execute ``pre-commit`` command."""
```

File: molecule/verifier/lint/precommit.py (hunk 28 above)
Comment (dmlb2000) [reply molecule/verifier/lint/precommit.py]: Unsure about this one, would you rather it say `Pre-commit tool verifier wrapper.` ???

**Comment by dmlb2000** [comment]
@decentral1se I'm unsure about this one, it would be nice to merge this.

File: molecule/verifier/lint/precommit.py (hunk 11 above)
Comment (ssbarnea) [reply molecule/verifier/lint/precommit.py]: @dmlb2000 Apparently when you created the PR you did not check the box to allow maintainers to update your PR. Please enable it so we can fix it and merge it.

**Comment by dmlb2000** [comment]
@ssbarnea I believe I've fixed everything discussed in the comments, and I'm not sure what button I'm supposed to push to enable edits. I don't have any branch protections on my fork of the repository anyone should be able to push what they need there to edit things.

**Comment by ssbarnea** [comment]
I am closing this because it was not updated in a very long time. Feel free to rebase it an raise it again once it passed.
